### PR TITLE
bpo-24928: Add test case for patch.dict using OrderedDict

### PR DESCRIPTION
**Pull request by eamanu**
[bpo-24928](https://bugs.python.org/issue24928): mock.patch.dict spoils order of items in collections.OrderedDict

This PR add test case for patch.dict using OrderedDict. 

This is just test, don't need NEWS.

Co-authored-by: Yu Tomita <nekobon@users.noreply.github.com>

https://bugs.python.org/issue24928

### CODE REVIEW

**Comment by ZackerySpytz** [comment]
Your PR is almost an exact copy of the Lib/unittest/test/testmock/testpatch.py changes
in https://bugs.python.org/file40488/issue24928.patch. That patch wasn't written
by you, so you should provide attribution to the original author, as stated in the developer's guide
(https://devguide.python.org/pullrequest/#converting-an-existing-patch-from-b-p-o-to-github).

File: Lib/unittest/test/testmock/testpatch.py
Comment (tirkarthi) [inline Lib/unittest/test/testmock/testpatch.py]: The assert looks incorrect that it's testing `list(foo) == sorted(foo)` but it should actually be testing whether `foo` contains `update_values` which on testing `foo['c']` gives `KeyError`. Sorry, I just noticed it now since the tests passed with original patch.

I checked this and `update_values` is a zip that becomes empty once it's iterated over and can be converted to a list to make sure that `update_values` can be used for subsequent calls. So how about below to make sure order is preserved?

```python
update_values = list(zip('cdefghijklmnopqrstuvwxyz', range(26)))
patched_items = list(foo.items()) + update_values

with patch.dict(foo, OrderedDict(update_values)):
    self.assertEqual(list(foo.items()), patched_items)
```

I would personally prefer a context manager here though other tests use `@patch.dict` with `test()` but that can wait for other reviewers.

**Comment by eamanu** [comment]
> Your PR is almost an exact copy of the Lib/unittest/test/testmock/testpatch.py changes
> in https://bugs.python.org/file40488/issue24928.patch. That patch wasn't written
> by you, so you should provide attribution to the original author, as stated in the developer's guide
> (https://devguide.python.org/pullrequest/#converting-an-existing-patch-from-b-p-o-to-github).

Thanks!. I added it to the PR description. In the next commit I will added it on commit message. Or is there other way to do it?

File: Lib/unittest/test/testmock/testpatch.py
Comment (eamanu) [reply Lib/unittest/test/testmock/testpatch.py]: Hi @tirkarthi 

Yes I checked it manually and the foo['c'] gives the KeyError. I will change for the list(zip()) that you say. And I will use context manager, I think that is more readable.

Maybe, we could think about support convert zip to dict on OrderedDict()?

Comment (tirkarthi) [reply Lib/unittest/test/testmock/testpatch.py]: > Maybe, we could think about support convert zip to dict on OrderedDict()?

Sorry, are you saying to use `patch.dict(foo, OrderedDict(dict(update_values)))` ? dict guarantees order currently and would help in catching the error if it changes in future. So I would prefer having `OrderedDict(update_values)`.

File: Lib/unittest/test/testmock/testpatch.py
Comment (tirkarthi) [inline Lib/unittest/test/testmock/testpatch.py]: I don't get why this was added. Any specific reason why this is needed in this test? Is there a case where this can fail perhaps due to `OrderedDict`?

Comment (eamanu) [reply Lib/unittest/test/testmock/testpatch.py]: I am just testing if the I modify the foo dict, this don't have to be modified out of the context

File: Lib/unittest/test/testmock/testpatch.py
Comment (eamanu) [reply Lib/unittest/test/testmock/testpatch.py]: I understand now.

File: Lib/unittest/test/testmock/testpatch.py
Comment (eamanu) [reply Lib/unittest/test/testmock/testpatch.py]: @tirkarthi Do you prefer that remove this test?

Comment (tirkarthi) [reply Lib/unittest/test/testmock/testpatch.py]: Yes, I think it's not related to the the test in question that is order of OrderedDict in patch.dict and more about a general one in whether being able to pop an element from a patched dict.

Comment (eamanu) [reply Lib/unittest/test/testmock/testpatch.py]: Ready

**Review by tirkarthi — APPROVED** [review]
LGTM. I will wait for review from someone else. Thanks for the PR.

File: Lib/unittest/test/testmock/testpatch.py
Comment (mariocj89) [inline Lib/unittest/test/testmock/testpatch.py]: Note this is also testing patch of ordered dicts (not just "patch dict"), Maybe just "test_patch_orderdict" or split the test in two.

**Comment by csabella** [comment]
@eamanu, please resolve the merge conflict and check the review by @mariocj89.  Thanks!

**Comment by csabella** [comment]
@eamanu, another reminder to please address the merge conflict and address the code review.  If there isn't any activity on this issue, then it will need to be closed.  Thank you!

**Comment by eamanu** [comment]
@csabella so sorry I didn't see the messages.

File: Lib/unittest/test/testmock/testpatch.py
Comment (eamanu) [reply Lib/unittest/test/testmock/testpatch.py]: Hi @mariocj89 I think this was resolved

**Comment by eamanu** [comment]
@csabella I beleave that this PR is finished.

**Comment by csabella** [comment]
@eamanu, thanks for updating the PR!

**Comment by tirkarthi** [comment]
I am fine with the test. @berkerpeksag had some opinions on the test in https://bugs.python.org/msg333890 so would wait for his approval on this.

**Pull request merged**
